x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/SearchableSnapshotIndexInputTests.java

```diff
@@ -0,0 +1,149 @@
+        }
+    }
+
+    public void testSeekOverflow() throws IOException {
+        for (int i = 0; i < 100; i++) {
+            byte[] input = randomUnicodeOfLength(randomIntBetween(1, 1000)).getBytes(StandardCharsets.UTF_8);
+            IndexInput indexInput = createIndexInput(input);
+            int firstReadLen = randomIntBetween(0, input.length - 1);
+            randomReadAndSlice(indexInput, firstReadLen);
+            expectThrows(IOException.class, () -> {
+                switch (randomIntBetween(0, 2)) {
+                    case 0:
+                        indexInput.seek(Integer.MAX_VALUE + 4L);
+                        break;
+                    case 1:
+                        indexInput.seek(-randomIntBetween(1, 10));
+                        break;
+                    default:
+                        int seek = input.length + randomIntBetween(1, 100);
+                        indexInput.seek(seek);
+                        break;
+                }
+            });
+        }
+    }
+
+    private byte[] randomReadAndSlice(IndexInput indexInput, int length) throws IOException {
+        int readPos = (int) indexInput.getFilePointer();
+        byte[] output = new byte[length];
+        while (readPos < length) {
+            switch (randomIntBetween(0, 3)) {
+                case 0:
+                    // Read by one byte at a time
+                    output[readPos++] = indexInput.readByte();
+                    break;
+                case 1:
+                    // Read several bytes into target
+                    int len = randomIntBetween(1, length - readPos);
+                    indexInput.readBytes(output, readPos, len);
+                    readPos += len;
+                    break;
+                case 2:
+                    // Read several bytes into 0-offset target
+                    len = randomIntBetween(1, length - readPos);
+                    byte[] temp = new byte[len];
+                    indexInput.readBytes(temp, 0, len);
+                    System.arraycopy(temp, 0, output, readPos, len);
+                    readPos += len;
+                    break;
+                case 3:
+                    // Read using slice
+                    len = randomIntBetween(1, length - readPos);
+                    IndexInput slice = indexInput.slice("slice (" + readPos + ", " + len + ") of " + indexInput.toString(), readPos, len);
+                    temp = randomReadAndSlice(slice, len);
+                    // assert that position in the original input didn't change
+                    assertEquals(readPos, indexInput.getFilePointer());
+                    System.arraycopy(temp, 0, output, readPos, len);
+                    readPos += len;
+                    indexInput.seek(readPos);
+                    assertEquals(readPos, indexInput.getFilePointer());
+                    break;
+                default:
+                    fail();
+            }
```

Comment:
> This is a really nice piece of fuzz testing

I do agree, that's why I shamelessly borrow them from `ByteArrayIndexInputTests`.

> Maybe we should also occasionally seek elsewhere in the stream, read a few bytes, then seek back here again.

The `testReadByte` `testReadBytes` tests in SearchableSnapshotDirectoryTests do random seeks, but only in a forward manner, as Lucene "tries" to do. I think  the coverage is good enough?

Reply:
I think we are missing coverage of reads that aren't purely forwards and contiguous, i.e. an interleaving of seeks and reads (particularly given the three different read methods). For instance, reading from more than one place in the file, possibly jumping back-and-forth across blobs, seems like an interesting workload for a cache. To be clear I don't think there's a bug here, just that if I was going to add a bug here in future then that's where I think I'd put it :grin:

Reply:
a2de0d2815e71b4de4cab624d4ff1ca43dc0d466 looks good, thanks, although we'd better not be using the [XKCD random number generator](https://xkcd.com/221/)...


![](https://imgs.xkcd.com/comics/random_number.png)

Reply:
That makes sense indeed. 

I added a change to the `randomReadAndSlice` method that seeks to a random position to read a single byte and then restore the previous position and pursue the execution (reading more than 1 byte would require more logic to handle corner cases though). It also overrides the byte read in the `output` value so that these extra reads are also verified as part of the test result.